config.md
# Runtime configuration

All component settings are stored in registry under `HKEY_LOCAL_MACHINE\Software\vmdisp9x` key.

Also plan create some configuration utility (beside) installer, but work on it has very low priority.

So every setting you needs to do thought registry editor (`regedit.exe`) or by insertion `*.reg` files by double click.

![Regedit](resource/docs/registry.png)

Setting tree is following:

```
HKEY_LOCAL_MACHINE\Software\vmdisp9x
  + apps
    + global = setting to all applications
       + hal = global setting for native DX/DD driver
       + mesa = global setting for Mesa
       + openglide = global setting for OpenGlide
       + wine = global setting for Wine
       - ddraw = global map to DD/DX7 runtime
       - d3d8 = global map to DX8 runtime
       - d3d9 = global map to DX9 runtime       
    + exe = setting to individual exe (identified by exe name)
      + <exe name>
         + hal = individual setting for HAL
         + mesa
         + openglide
         + wine
         - ddraw = per application DD/DX7 runtime selection
         - d3d8
         - d3d9
    + profile = complex profiles (identified by path with regex)
      + <profile name>
         - path = regex of path to exe
         - name = profile name
         - desc = short description
         + hal
         + mesa
         + openglide
         + wine
         - ddraw
         - d3d8
         - d3d9
  + driver = universal display settings
  + svga = display settings related to VMware SVGA-II gpu
  + vesa = display settings related to VESA driver
```

NOTE: you can enter numeric values as DWORD or as string, it'll be converted automatically.

## Universal display settings

| Name              | Default value | Description |
| ----------------- | -----------   | ----------- |
| FORCE_SOFTWARE    | 0             | 1 to use software OpenGL even if there is HW accelerated |
| FORCE_QEMU3DFX    | 0             | 1 when you has working QEMU-3Dfx wrapper |
| REV_QEMU3DFX      | (empty)       | tag to override driver default (build-in) |
| GAMMA_QUIRK       | 0             | 0 - auto detect gamma range |
|                   |               | 1 - gamma in range 0 - 127 |
|                   |               | 2 - gamma in range 0 - 255 |
| OVERLAY_n_SIZE    | 0             | define region on VRAM that can be used as frame buffer, but without notify the system - for example simulate relay on Voodoo cards. |

## SVGA driver

| Name              | Default value | Description |
| ----------------- | -----------   | ----------- |
| HWCursor          | 0             | 1 enable HW cursor, 0 draw cursor by software |
| VRAMLimit         | 128           | limits maximum of VRAM (for Win9x maximum around 256) | 
| CommandBuffers    | 1             | enable newer style commands (on vGPU10 must be enabled) |
| PreferFIFO        | 1             | prefer FIFO over command buffer when is possible |
| RGB565bug         | 0             | fix VB 6.1 bug when RGB565 formats has wrong flags |
| HWVersion         | 2             | 2 = SVGA_VERSION_2  |
| NoMultisample     | 1             | disable multisample (vmware bug) |
| RegMultisample    | 0             | value to SVGA_REG_MSHINT register |
| AsyncMOBs         | 1             | number of MOBs to be created simulately |
| NoScreenAccel     | 0             | when enable, all frame buffer operation will be done strictly by CPU |

## VESA driver

| Name | Default value | Description |
| ---- | ------------- | ----------- |
| VRAMLimit | 128      | Maximum video RAM (in MB) exposed to system, allow to use adapters when have more then 256 MB VRAM, VRAM is usually utilized only by frame buffers, 32 MB is usually enough for FullHD resolution (1920 x 1080).
| MTRR | 1             | Allow to use Intel MTRR registry to accelerate CPU to VRAM operations |
| DosWindowSetMode | 0 | Allow to DOS application to call set mode when DOS application is in window mode |
| HWDoubleBuffer | 2 | 0 - double buffering is fully emulated, 
| | | 1 - allow frame buffer swap (if supported),
| | | 2 - allow frame buffer swap during vertical retrace (if supported).
| NoMemTest | 0 | when 1 disable video memory testing. Some GPUs may have some I/O in linear frame buffer space and overwrite this memory can confuse the device. |


### DosWindowSetMode

[More description in VMDisp9x readme](https://github.com/JHRobotics/vmdisp9x?tab=readme-ov-file#dos-in-window)

## HAL

| Name              | Per application | Default value | Description |
| ----------------- | --------------- | -----------   | ----------- |
| ddi               | **global only** | 8             | set DDI level, values can be 0, 3, 5, 6. 7, 8 |
| hwtl              | **global only** | 8             | enable HW T&L in ddi level: 8 - HW T&L in DX 9, 8, 7; 7 - HW T&L in DX 7 only; 0 - HW T&L disabled |
|forceos            | **global only** | 0             | force to use software rasterisation |
| sysmem            | **global only** | 0             | force to allocate surfaces in RAM |
| vertexblend       | **global only** | 0             | allow to emulate DX vertex blend (slow and buggy) |
| palette           | **global only** | 0             | allow to use indexed palette textures (buggy in some games) |
| readback          | yes             | 0             | when 1, render output is read back to  surface, slow, but keep for compatibility |
| touchdepth        | yes             | 0             | when 1, depth buffer is read and write from and to surface, slow, only need when application needs manipulate with depth buffer |
| filter_bug        | yes             | auto          | SVGA driver has bug when filtering isn't applied correctly |
| s3tc_bug          | yes             | 1             | software conversion of S3 style compressed textures |
| lowdetail         | yes             | 0             | reduce detail, from 0 (no detail reduction) to 3 (lowest detail) |

### lowdetail

This setting allow turn off some graphical feature to faster rendering. Note: this setting affect only software 3D, respective detail reduction will be visible on HW and SW renderer but only on SW renderer affect performance. 

| Level | Description |
| ----- | ----------- |
| 0     | All graphical features on |
| 1     | Disabled dithering and edge antialiasing |
| 2     | level 1 + texture filtering is forced to nearest neighbour including mipmaps |
| 3     | level 2 + shading is forced to flat |

I recommend to use Level 1 every time when SW rendering is ON. On Level 2 games usually looks like old games when switched to software renderer. Level 3 has visible artefact when lighting is used. When you need more FPS from SW renderer, please reduce resolution, you can also try to turn off HW T&L (set `hwtl` to `0`), SW T&L done in DX runtime is paradoxically faster that than HW T&L performed by SW renderer.


## MESA

These keys have same names and behaviour like [Mesa Environment Variables](https://docs.mesa3d.org/envvars.html), so there are few important values

| Name | Default | Description |
| ---- | ------- | ----------- |
| LP_NATIVE_VECTOR_WIDTH | 128 | CPU register size for LLVM pipe renderer, 128 = SSE, or 256 = AVX, when you have AVX support in CPU and enabled support in system, you can set this variable to 256 for faster software rendering |
| MESA_EXTENSION_MAX_YEAR | - | The string with OpenGL extensions too long to cause buffer overflow in old games (Q2 and Q3 engines games). When enter year, the newer extension is not listed (but still works). Usually good values are '2004' or '2006'. Because HAL and WINE internally using Mesa, please do not set this setting globally.|


## OpenGlide

| Name | Default | Description |
| ---- | ------- | ----------- |
| Disabled | 0 |
| FogEnable | 0 |
| InitFullScreen | 0 |
| Textures32bit | 0 |
| HideCursor | 1 |
| NoSplash | 0 |
| SwapInterval | 0 |
| Resolution | 1 |
| NumTMU | 1 |
| SSTType | 1 |
| TextureMemorySize | 4 |
| FrameBufferMemorySize | 4 |


## WINE



"VideoMemorySize"="256"

## Examples


changelog.txt
0.8.2025.52
! VESA display driver
! VBOX: page fault in VBox VGA driver
- small stability improvements
- fixed corrupted driver inf in some cases

0.8.2025.51
- VMware: VRAM test (VMware bug)
- QEMU: vmware adapter compatibility fixes
- SVGA: keep line align when autoresize/in custom modes
- HAL: don't call memory helpers from ddraw.dll
- HAL: better surface tracking

0.8.2025.50
! DirectX native driver
! no DLL replacement
- HAL: DDI 8
- Wine fully optional
- SVGA: stable memory allocation
- MESA: revert to Mesa 23.1.9 as most stable
- Wine: original DLL are now hot patched (allow to use without DLL replace)

0.6.2024.39
! stability fixes
! double buffering support (or emulation)
! Hires <-> VGA mode switching correctly
- Gamma support (disabled by default)
- Driver: HW cursor in 8/16/32 bpp (VMware)
- Increased compatibility list

0.6.2024.38
! no hard wired modes on DDHAL (1280x800 didn't work for example)
! no unknown control to DIP (causes BSOD)
! Wine: broken by LTO (introduced in 0.6.2024.37)

0.6.2024.37
! native DDHAL now operable (introduced in 0.6.2024.36)
! reduced wasteful surface align

0.6.2024.36
! fixed missing non English chars
! VMWare vGPU10 support
! fixed memory waste when allocating large regions
! better handling out-of-memory errors
! profiles for different hypervisors
- Better HW detection
- Driver: fixed race-conditional bugs
- Driver: SVGA screen target support
- Driver: Smooth font support
- Driver: driver version reporting to HV
- Driver: better VMware support
- Mesa/SVGA: better memory limitation
- Mesa/SVGA: screen target support (vGPU10, VBox only)
- Mesa/SVGA: OpenGL 4.1 (vGPU10, VBox), 4.3 (vGPU10, VMware)
- Mesa/SVGA: better compatibility detection
- Mesa/LLVMpipe: LLVM 18.1.8
- Mesa/LLVMpipe: OpenGL 4.5
- Mesa/Nine: fixed diffuse lights
- Mesa: faster readpixels
- Wine: S3TC support (by @hpvb)
- Wine: explorer, iexplore blacklisted
- Increased compatibility list

0.5.2024.29
! Wine works on Windows 95
! OpenGlide works with MiniGL wrappers
- Driver: VIDEOPARAMETERS support
- SVGA/Mesa: added GMR cache (require 1024 MB RAM)
- Mesa: fixed page fault with llvmpipe/softpipe
- Wine: fixed creating shader object (even if no shader support)
- Wine: better restore screen after game/app quit
- OpenGlide: disabled fog (and problems with it)
- OpenGlide: restore screen after quit
- OpenGlide: SSE and MMX optimized format conversion
- Increased compatibility list

0.5.2024.24
! fixed the most of mouse cursor bugs
! limited RAM memory usage
! fixed some setup GUI bugs (introduced in 0.4.2023.18)
- vGPU10 finaly works
- DirectDraw HAL (2D only)
- Driver: reduced usage of 16bit RING-3 code (increase stability and speed)
- Driver: command buffers disabled for vGPU9 deviced (FIFO usage looks stable and faster)
- Wine: replaced window proc daisy chain with hooks (increase compatibility by a lot)
- Wine: report no shader support for vGPU9 devices (they didn't work anyway)
- Wine: DX 8 + D9 implementation switcher
- Wine: DLL version is set to 4.9.0.904
- OpenGlide: fixed multi-thread rendering
- Mesa: optimized driver
- Mesa: various rendering fixes
- Mesa: Nine support (experimental)
- Mesa: 8-to-9 wrapper for Nine (experimental)
- Mesa: vGPU10 - OpenGL version 3.3
- QEMU: QEMU-3Dfx OpenGL ICD driver

0.4.2023.19
! fixed GPUs without CB (VBox <= 6.1)
! fixed GMR allocation pool (introduced in 0.4.2023.13)
! enabled direct rendering (introduced in 0.4.2023.18)
! fixed direct rendering in window mode
- optimized locks
- driver configuration in registry
- specific tune for VirtualBox/VMware

0.4.2023.18
! fixed some stability issues in CRT (malloc, strndup, ...)
! More stability on non-SSE guests (introduced in 0.4.2023.13)
- Virtual GPU gen. 10 supported!
- Mesa: 23.1.6
- Mesa: faster rendering
- glchecker: fixed startup crash
- glchecker: fixed missing textures on OpenGL 2.x
- Wine: some stability issues
- Wine: support for GPU gen 10
- Driver: SVGA - HW cursor is supported and enabled (32 bpp only)

settings.c
/******************************************************************************
 * Copyright (c) 2024 Jaroslav Hensl                                          *
 *                                                                            *
 * Permission is hereby granted, free of charge, to any person                *
 * obtaining a copy of this software and associated documentation             *
 * files (the "Software"), to deal in the Software without                    *
 * restriction, including without limitation the rights to use,               *
 * copy, modify, merge, publish, distribute, sublicense, and/or sell          *
 * copies of the Software, and to permit persons to whom the                  *
 * Software is furnished to do so, subject to the following                   *
 * conditions:                                                                *
 *                                                                            *
 * The above copyright notice and this permission notice shall be             *
 * included in all copies or substantial portions of the Software.            *
 *                                                                            *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,            *
 * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES            *
 * OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND                   *
 * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT                *
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,               *
 * WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING               *
 * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR              *
 * OTHER DEALINGS IN THE SOFTWARE.                                            *
 *                                                                            *
*******************************************************************************/
#include <windows.h>
#include <Shlobj.h>
#include <stdio.h>
#include <stdint.h>
#include <math.h>

#include "softgpu.h"

#include "nocrt.h"

#define ST_DEF 0
#define VRAM_DEF 128
#define SVGA_MEM_MAX 400
#define DEFAULT_INST_PATH "C:\\drivers\\softgpu"

#define T_CHECKBOX  1
#define T_RADIO     2
#define T_INPUT_STR 3
#define T_INPUT_NUM 4
#define T_DROPDOWN  5

typedef struct _settings_item_t
{
	DWORD menu;
	DWORD type;
	DWORD pos;
	const char *name;
	DWORD dvalue;
	const char *svalue;
	char *svalue_freeptr;
	int group;
	int disabled;
} settings_item_t;

static const settings_item_t settings_def[] = 
{
	{CHBX_MSCRT,           T_CHECKBOX,   0, "mscrt",    0, NULL, NULL, 0, 0},
	{CHBX_DX,              T_CHECKBOX,   1, "dxredist", 1, NULL, NULL, 0, 0},
	{RAD_NORMAL,           T_RADIO,      2, "bin_mmx",  1, NULL, NULL, 1, 0},
	{RAD_SSE,              T_RADIO,      3, "bin_sse",  0, NULL, NULL, 1, 0},
	{RAD_SSE4,             T_RADIO,      4, "bin_sse4", 0, NULL, NULL, 1, 0},
	{CHBX_GL95,            T_CHECKBOX,   5, "gl95",     0, NULL, NULL, 0, 0},
	{CHBX_WINE,            T_CHECKBOX,   6, "wine",     1, NULL, NULL, 0, 0},
	{CHBX_GLIDE,           T_CHECKBOX,   7, "glide",    1, NULL, NULL, 0, 0},
	{CHBX_SIMD95,          T_CHECKBOX,   8, "simd95",   0, NULL, NULL, 0, 0},
	{CHBX_3DFX,            T_CHECKBOX,   9, "3dfx",     1, NULL, NULL, 0, 0},
	{CHBX_FIXES,           T_CHECKBOX,  10, "fixes",    1, NULL, NULL, 0, 0},
	{CHBX_QXGA,            T_CHECKBOX,  11, "res_qxga", 0, NULL, NULL, 0, 0},
	{CHBX_1440,            T_CHECKBOX,  12, "res_1440", 0, NULL, NULL, 0, 0},
	{CHBX_4K,              T_CHECKBOX,  13, "res_2160", 0, NULL, NULL, 0, 0},
	{CHBX_5K,              T_CHECKBOX,  14, "res_2880", 0, NULL, NULL, 0, 0},
	{RAD_DD_HAL,           T_RADIO,     15, "dd_hal",   1, NULL, NULL, 2, 0},
	{RAD_DD_WINE,          T_RADIO,     16, "dd_wine",  0, NULL, NULL, 2, 0},
	{RAD_D8_HAL,           T_RADIO,     45, "d8_hal",   0, NULL, NULL, 3, 0},
	{RAD_D8_WINE,          T_RADIO,     17, "d8_wine",  1, NULL, NULL, 3, 0},
	{RAD_D8_NINE,          T_RADIO,     18, "d8_nine",  0, NULL, NULL, 3, 0},
	{RAD_D9_HAL,           T_RADIO,     46, "d9_hal",   0, NULL, NULL, 4, 0},
	{RAD_D9_WINE,          T_RADIO,     19, "d9_wine",  1, NULL, NULL, 4, 0},
	{RAD_D9_NINE,          T_RADIO,     20, "d9_nine",  0, NULL, NULL, 4, 0},
	{CHBX_BUG_PREFER_FIFO, T_CHECKBOX,  21, "fifo",     1, NULL, NULL, 0, 0},
	{CHBX_BUG_RGB565,      T_CHECKBOX,  22, "rgb565",   1, NULL, NULL, 0, 0},
	{CHBX_BUG_DX_FLAGS,    T_CHECKBOX,  23, "dxflags",  0, NULL, NULL, 0, 0},
	{LBX_PROFILE,          T_DROPDOWN,  24, "profile",  0, NULL, NULL, 0, 0},
	{INP_PATH,             T_INPUT_STR, 25, "path",     0, DEFAULT_INST_PATH, NULL, 0, 0},
	{INP_VRAM_LIMIT,       T_INPUT_NUM, 26, "vram",     VRAM_DEF, NULL, NULL, 0, 0},
	{CHBX_DOTCOM,          T_CHECKBOX,  31, "dotcom",   0, NULL, NULL, 0, 0},
	{CHBX_WS2,             T_CHECKBOX,  32, "ws2",      0, NULL, NULL, 0, 0},
	{CHBX_NO_MULTISAMPLE,  T_CHECKBOX,  36, "no_multisample",  0, NULL, NULL, 0, 0},
	{INP_ASYNCMOBS,        T_INPUT_NUM, 37, "asyncmobs", 1, NULL, NULL, 0, 0},
	{CHBX_GMR_CACHE,       T_CHECKBOX,  38, "gmrcache",  0, NULL, NULL, 0, 0},
	{IMP_SVGA_MEM_MAX,     T_INPUT_NUM, 39, "svgamemmax", SVGA_MEM_MAX, NULL, NULL, 0, 0},
	{CHBX_SW_GAMMA,        T_CHECKBOX,  40, "swgamma",   0, NULL, NULL, 0, 0},
	{CHBX_DMA_TO_FB,       T_CHECKBOX,  41, "dmatofb",   0, NULL, NULL, 0, 0},
	{CHBX_HWCURSOR,        T_CHECKBOX,  42, "hwcursor",  0, NULL, NULL, 0, 0},
	{CHBX_MESA_ALT,        T_CHECKBOX,  44, "mesa_alt",  0, NULL, NULL, 0, 0},
	{CHBX_NO_INSTALL,      T_CHECKBOX,  47, "no_install",0, NULL, NULL, 0, 0},
	{RAD_LOWDETAIL_0,      T_RADIO,     48, "no_lowdetail",    1, NULL, NULL, 5, 0},
	{RAD_LOWDETAIL_1,      T_RADIO,     49, "lowdetail_lev1",  0, NULL, NULL, 5, 0},
	{RAD_LOWDETAIL_2,      T_RADIO,     50, "lowdetail_lev2",  0, NULL, NULL, 5, 0},
	{RAD_LOWDETAIL_3,      T_RADIO,     51, "lowdetail_lev3",  0, NULL, NULL, 5, 0},
	{0,                    0,            0, NULL,        0, NULL, NULL, 0, 0}
};

static settings_item_t settings_cur[sizeof(settings_def)/sizeof(settings_item_t)];

static BOOL settings_cur_init = FALSE;

static void settingClearGroup(int group)
{
	settings_item_t *item;
	
	if(group == 0)
		return;
	
	for(item = &settings_cur[0]; item->menu != 0; item++)
	{
		if(item->group == group)
		{
			item->dvalue = 0;
		}
	}
}

void settingsApply(HWND hwnd)
{
	settings_item_t *item;
	
	for(item = &settings_cur[0]; item->menu != 0; item++)
	{
		HWND hItem = GetDlgItem(hwnd, item->menu);
		if(hItem)
		{
			switch(item->type)
			{
				case T_RADIO:
				case T_CHECKBOX:
					if(item->disabled)
					{
						CheckDlgButton(hwnd, item->menu, BST_UNCHECKED);
						LONG sty = GetWindowLongA(hItem, GWL_STYLE) | WS_DISABLED;
						SetWindowLongA(hItem, GWL_STYLE, sty);
					}
					else
					{
						LONG sty = GetWindowLongA(hItem, GWL_STYLE) & (~WS_DISABLED);
						SetWindowLongA(hItem, GWL_STYLE, sty);
						
						if(item->dvalue)
							CheckDlgButton(hwnd, item->menu, BST_CHECKED);
						else
							CheckDlgButton(hwnd, item->menu, BST_UNCHECKED);
					}
					break;
				case T_INPUT_NUM:
				{
					char inp[64];
					sprintf(inp, "%u", item->dvalue);					
					SetWindowTextA(hItem, inp);
					break;
				}
				case T_INPUT_STR:
					SetWindowTextA(hItem, item->svalue);
					break;
				case T_DROPDOWN:
					SendMessage(hItem, (UINT)CB_SETCURSEL, (WPARAM)item->dvalue, (LPARAM)0);
					break;
			}
		}
	}
}

void settingsReadback(HWND hwnd)
{
	settings_item_t *item;
	
	for(item = &settings_cur[0]; item->menu != 0; item++)
	{
		DWORD value;
		HWND hItem = GetDlgItem(hwnd, item->menu);
		if(hItem)
		{
			switch(item->type)
			{
				case T_RADIO:
					value = IsDlgButtonChecked(hwnd, item->menu) ? 1 : 0;
					if(value > 0)
					{
						settingClearGroup(item->group);
						item->dvalue = value;
					}
					break;
				case T_CHECKBOX:
					value = IsDlgButtonChecked(hwnd, item->menu) ? 1 : 0;
					item->dvalue = value;
					break;
				case T_INPUT_NUM:
				{
					char inp[64];
					if(GetWindowTextA(hItem, inp, 64) != 0)
					{
						item->dvalue = strtol(inp, NULL, 0);
					}
					break;
				}
				case T_INPUT_STR:
				{
					if(item->svalue_freeptr != NULL)
					{
						free(item->svalue_freeptr);
						item->svalue_freeptr = NULL;
						item->svalue = NULL;
					}
					
					size_t s = GetWindowTextLengthA(hItem) + 1;
					item->svalue_freeptr = malloc(s);
					if(item->svalue_freeptr)
					{
						GetWindowTextA(hItem, item->svalue_freeptr, s);
						item->svalue = item->svalue_freeptr;
					}
					break;
				}
				case T_DROPDOWN:
					item->dvalue = SendMessage(hItem, (UINT)CB_GETCURSEL, (WPARAM)0, (LPARAM)0);
					break;
			}
		}
	}
}

void settingsApplyProfile(int profile)
{
	settings_item_t *item;
	
	for(item = &settings_cur[0]; item->menu != 0; item++)
	{
		const char *v = iniSectionsValue("[profile]", profile, item->name);
		if(v == NULL)
		{
			v = iniValue("[defaults]", item->name);
		}
		
		if(v != NULL)
		{
			DWORD dw = strtoul(v, NULL, 0);
			settingsSet(item->menu, dw);
		}
	}
}

static BOOL have_3dfx_files()
{
	BOOL r = FALSE;
	const char *drv_voodoo = iniValue("[softgpu]", "voodoo2path");
	if(drv_voodoo != NULL)
	{
		if(is_dir(drv_voodoo))
		{
			char testpath[MAX_PATH];
			strcpy(testpath, drv_voodoo);
			strcat(testpath, "\\3dfxSpl2.dll");
			if(is_file(testpath))
			{
				strcpy(testpath, drv_voodoo);
				strcat(testpath, "\\3dfxSpl3.dll");
				if(is_file(testpath))
				{
					strcpy(testpath, drv_voodoo);
					strcat(testpath, "\\3dfxVGL.dll");
					if(is_file(testpath))
					{
						r = TRUE;
					}
				}
			}
		}
	}
	
	return r;
}

static void settingsCompute()
{
	if(version_compare(&sysver, &WINVER98) < 0)
	{
		if(!hasOpengl)
		{
			settingsSetByName("gl95", 1);
		}
		
		if(!hasOle32)
		{
			settingsSetByName("dotcom", 1);
		}
		
		if(!hasWS2)
		{
			settingsSetByName("ws2", 1);
		}
	}
	else
	{
		settingsDisableByName("gl95",   1);
		settingsDisableByName("dotcom", 1);
		settingsDisableByName("ws2",    1);
	}
	
	if(!isNT)
	{
		if(!hasCRT)
		{
			settingsSetByName("mscrt", 1);
		}
		
		if(version_compare(&dxver, &dxtarget) < 0 || reinstall_dx)
		{
			settingsSetByName("dxredist", 1);
		}
	}
	else
	{
		settingsDisableByName("mscrt", 1);
		settingsDisableByName("dxredist", 1);
	}
	
	const char *drv_sse4 = iniValue("[softgpu]", "drvpath.sse4");
	BOOL sse4_bins = drv_sse4 != NULL && is_dir(drv_sse4);
	
	if(sse4_bins)
	{
		if(version_compare(&sysver, &WINVER98) >= 0 && hasSSE42)
		{
			settingsSetByName("bin_sse4", 1);
		}
		else if(version_compare(&sysver, &WINVER98) >= 0 && hasSSE3)
		{
			settingsSetByName("bin_sse", 1);
		}
		else
		{
			settingsSetByName("bin_mmx", 1);
		}
	}
	else
	{
		settingsDisableByName("bin_sse4", 1);
		
		if(version_compare(&sysver, &WINVER98) >= 0 && hasSSE3)
		{
			settingsSetByName("bin_sse", 1);
		}
		else
		{
			settingsSetByName("bin_mmx", 1);
		}
	}
	
	if(version_compare(&sysver, &WINVERME) >= 0)
	{
		settingsDisableByName("simd95", 1);
	}
	
	if(!have_3dfx_files())
	{
		settingsDisableByName("3dfx", 1);
	}
}

void settingsSet(DWORD menu, DWORD value)
{
	settings_item_t *item;
	
	for(item = &settings_cur[0]; item->menu != 0; item++)
	{
		if(item->menu == menu)
		{
			switch(item->type)
			{
				case T_RADIO:
					if(value > 0)
					{
						settingClearGroup(item->group);
						item->dvalue = value;
					}
					break;
				case T_CHECKBOX:
				case T_INPUT_NUM:
				case T_DROPDOWN:
					item->dvalue = value;
					break;
			}
			break;
		}
	}
}

void settingsSetByName(const char *name, DWORD value)
{
	settings_item_t *item;
	
	for(item = &settings_cur[0]; item->menu != 0; item++)
	{
		if(strcmp(name, item->name) == 0)
		{
			switch(item->type)
			{
				case T_RADIO:
					if(value > 0)
					{
						settingClearGroup(item->group);
						item->dvalue = value;
					}
					break;
				case T_CHECKBOX:
				case T_INPUT_NUM:
				case T_DROPDOWN:
					item->dvalue = value;
					break;
			}
			break;
		}
	}
}

void settingsDisableByName(const char *name, int disabled)
{
	settings_item_t *item;
	
	for(item = &settings_cur[0]; item->menu != 0; item++)
	{
		if(strcmp(name, item->name) == 0)
		{
			item->disabled = disabled;
			break;
		}
	}
}

/* free allocated strings */
void settingsFree()
{
	settings_item_t *item;
	
	for(item = &settings_cur[0]; item->menu != 0; item++)
	{
		if(item->type == T_INPUT_STR)
		{
			if(item->svalue_freeptr != NULL)
			{
				free(item->svalue_freeptr);
				item->svalue_freeptr = NULL;
				item->svalue = NULL;
			}
		}
	}
}

/* from last save something can change */
static void settingsReEval()
{	
	if(hasOpengl)
	{
		settingsSetByName("gl95", 0);
	}
		
	if(hasOle32)
	{
		settingsSetByName("dotcom", 0);
	}
		
	if(hasWS2)
	{
		settingsSetByName("ws2", 0);
	}
	
	if(hasCRT)
	{
		settingsSetByName("mscrt", 0);
	}
		
	if(version_compare(&dxver, &dxtarget) >= 0 && reinstall_dx == FALSE)
	{
		settingsSetByName("dxredist", 0);
	}
}

/* load default settings */
void settingsReset()
{
	if(settings_cur_init)
	{
		settingsFree();
	}
	
	memcpy(&settings_cur[0], &settings_def[0], sizeof(settings_def));
	
	settingsCompute();
	
	if(get_gpu_profile() >= 0)
	{
		DWORD p = get_gpu_profile();
		settingsSetByName("profile", p);
		settingsApplyProfile(p);
	}
	else
	{
		const char *defProfile = iniValue("[softgpu]", "default_profile");
		if(defProfile != NULL)
		{
			DWORD p = strtoul(defProfile, NULL, 0);
			settingsSetByName("profile", p);
			settingsApplyProfile(p);
			//printf("profile: %u\n", p);
		}
		else
		{
			settingsApplyProfile(0);
		}
	}
	
	const char *defPath = iniValue("[softgpu]", "default_path");
	if(defPath != NULL)
	{
		settingsSetStr(INP_PATH, defPath, FALSE);
	}
	
	/* read from registry first time */
	if(!settings_cur_init)
	{
		readSettings();
		settingsReEval();
	}
	
	settings_cur_init = TRUE;
}

BOOL isSettingSet(DWORD menu)
{
	settings_item_t *item;
	for(item = &settings_cur[0]; item->menu != 0; item++)
	{
		if(item->menu == menu)
		{
			return item->dvalue;
		}
	}
	
	return FALSE;
}

DWORD settingReadDW(DWORD menu)
{
	settings_item_t *item;
	for(item = &settings_cur[0]; item->menu != 0; item++)
	{
		if(item->menu == menu)
		{
			switch(item->type)
			{
				case T_CHECKBOX:
				case T_RADIO:
				case T_INPUT_NUM:
				case T_DROPDOWN:
					return item->dvalue;
					break;
				case T_INPUT_STR:
					return strtoul(item->svalue, NULL, 0);
					break;
			}
		}
	}
	
	return 0;
}

const char *settingsReadStr(DWORD menu)
{
	settings_item_t *item;
	for(item = &settings_cur[0]; item->menu != 0; item++)
	{
		if(item->menu == menu)
		{
			switch(item->type)
			{
				case T_INPUT_STR:
					return item->svalue;
			}
		}
	}
	
	return NULL;
}

void settingsSetStr(DWORD menu, const char *str, BOOL copy)
{
	settings_item_t *item;
	for(item = &settings_cur[0]; item->menu != 0; item++)
	{
		if(item->menu == menu)
		{
			switch(item->type)
			{
				case T_INPUT_STR:
					if(item->svalue_freeptr != NULL)
					{
						free(item->svalue_freeptr);
						item->svalue_freeptr = NULL;
						item->svalue = NULL;
					}
					
					if(!copy)
					{
						item->svalue = str;
					}
					else
					{
						size_t s = strlen(str) + 1;
						item->svalue_freeptr = malloc(s);
						if(item->svalue_freeptr)
						{
							memcpy(item->svalue_freeptr, str, s);
							item->svalue = item->svalue_freeptr;
						}
					}
					break;
			}
		}
	}
}

static DWORD getSettingsLevel()
{
	DWORD settings_level = atoi(iniValueDef("[softgpu]", "settings_level", "0"));
	if(settings_level == 0)
	{
		settings_level = SOFTGPU_BUILD;
	}
	
	return settings_level;
}

#define MAX_CONF_DW 4

#define KEYNAME_FMT "HKCU\\SOFTWARE\\SoftGPU\\setup-%d\\%s"

void writeSettings()
{
	char keyname[MAX_PATH];
	DWORD confdw[MAX_CONF_DW];
	
	DWORD settings_level = getSettingsLevel();
	
	memset(&confdw[0], 0, sizeof(confdw));
	
	settings_item_t *item;
	for(item = &settings_cur[0]; item->menu != 0; item++)
	{
		switch(item->type)
		{
			case T_CHECKBOX:
			case T_RADIO:
				DWORD p = item->pos / 32;
				DWORD i = item->pos % 32;
				if(item->dvalue > 0)
				{
					confdw[p] |= (1 << i);
				}
				break;
			case T_INPUT_NUM:
			case T_DROPDOWN:
				sprintf(keyname, KEYNAME_FMT, settings_level, item->name);
				registryWriteDWORD(keyname, item->dvalue);
				break;
			case T_INPUT_STR:
				sprintf(keyname, KEYNAME_FMT, (item->menu == INP_PATH) ? SOFTGPU_BUILD : settings_level, item->name);
				registryWrite(keyname, item->svalue, WINREG_STR);
				break;
		}
	}
	
	int j;
	for(j = 0; j < MAX_CONF_DW; j++)
	{
		sprintf(keyname, KEYNAME_FMT "%02d", settings_level, "cfg", j);
		registryWriteDWORD(keyname, confdw[j]);
	}
}

BOOL readSettings()
{
	char keyname[MAX_PATH];
	char value[MAX_PATH];
	DWORD confdw[MAX_CONF_DW];
	
	DWORD settings_level = getSettingsLevel();
	
	int j;
	for(j = 0; j < MAX_CONF_DW; j++)
	{
		sprintf(keyname, KEYNAME_FMT "%02d", settings_level, "cfg", j);
		if(!registryReadDWORD(keyname, &confdw[j]))
		{
			return FALSE;
		}
	}
	
	settings_item_t *item;
	for(item = &settings_cur[0]; item->menu != 0; item++)
	{
		switch(item->type)
		{
			case T_CHECKBOX:
			case T_RADIO:
				DWORD p = item->pos / 32;
				DWORD i = item->pos % 32;
				item->dvalue = (confdw[p] >> i) & 0x1;
				break;
			case T_INPUT_NUM:
			case T_DROPDOWN:
				sprintf(keyname, KEYNAME_FMT, settings_level, item->name);
				registryReadDWORD(keyname, &item->dvalue);
				break;
			case T_INPUT_STR:
				sprintf(keyname, KEYNAME_FMT, (item->menu == INP_PATH) ? SOFTGPU_BUILD : settings_level, item->name);
				if(registryRead(keyname, value, MAX_PATH))
				{
					if(item->svalue_freeptr != NULL)
					{
						free(item->svalue_freeptr);
						item->svalue_freeptr = NULL;
						item->svalue = NULL;
					}
					
					size_t s = strlen(value)+1;
					item->svalue_freeptr = malloc(s);
					if(item->svalue_freeptr)
					{
						memcpy(item->svalue_freeptr, value, s);
						item->svalue = item->svalue_freeptr;
					}
				}
				break;
		}
	}
	
	return TRUE;
}
